pkg/cdi/cache.go

```diff
@@ -0,0 +1,551 @@
+package cdi
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+
+	oci "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/generate"
+	"github.com/pkg/errors"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	// DefaultStaticDir is the default directory for static CDI Specs.
+	DefaultStaticDir = "/etc/cdi"
+	// DefaultDynamicDir is the default directory for dynamically generated CDI Specs.
+	DefaultDynamicDir = "/var/run/cdi"
+)
+
+var (
+	// DefaultSpecDirs contains the default CDI Spec directories.
+	DefaultSpecDirs = []string{DefaultStaticDir, DefaultDynamicDir}
```

Comment:
Probably not 100 % mandatory to export the slice (if that's what you mean) since not setting the related option will cause the package to use these defaults. Is your concern that someone can alter the defaults since it is visible outside the package ?

Reply:
Yes, I wanted to avoid having users manipulate these because they can.

Reply:
A bit nitpicking on my side and of course we can make that variable unexported without any problems. That said, isn't it quite a natural abstraction for changing the default set of CDI directories to, well do just that: change the default set of CDI directories ?

Reply:
It may be personal preference, but I prefer to have this exposed as arguments to a constructor / `New*` function with an explicit name for the default case. This also makes testing easier since we don't have to manipulate global state to get tests to work the way we want.

Reply:
It already works that way... (using the `cdi.WithSpecDirs(cdi.DefaultSpecDirectories...)` option). This is exported exactly for that reason, so that it is possible to feed explicitly in the default options, instead of hoping that omitting the right options is going to do exactly that if, for whatever reason you need to do that. Plus so that clients can monitor CDI Spec directories if they wish and do intelligent on-demand Refresh() if they want to. Without this being exported (one way or another, there are other ways for that to work, too), they would not know what directories to monitor.